packages/image-block/src/ImageBlock.tsx

```diff
@@ -1,21 +1,72 @@
+import { Asset, useBlockAssets, useBlockSettings, useEditorState } from '@frontify/app-bridge';
 import '@frontify/fondue-tokens/styles';
 import { BlockProps } from '@frontify/guideline-blocks-settings';
 import 'tailwindcss/tailwind.css';
 import { Settings, mapCaptionPositionClasses } from './types';
 import { UploadPlaceholder } from './components/UploadPlaceholder';
 import { ImageCaption } from './components/ImageCaption';
-import { IMAGE_SETTING_ID } from './settings';
+import { ATTACHMENTS_SETTING_ID, IMAGE_SETTING_ID } from './settings';
 import { joinClassNames } from '@frontify/guideline-blocks-shared';
 import { Image } from './components/Image';
 
 export const ImageBlock = ({ appBridge }: BlockProps) => {
     const [blockSettings, setBlockSettings] = useBlockSettings<Settings>(appBridge);
-    const { blockAssets } = useBlockAssets(appBridge);
+    const { blockAssets, updateAssetIdsFromKey, addAssetIdsToKey, deleteAssetIdsFromKey } = useBlockAssets(appBridge);
     const image = blockAssets?.[IMAGE_SETTING_ID]?.[0];
+    const attachments = blockAssets?.[ATTACHMENTS_SETTING_ID] || [];
     const isEditing = useEditorState(appBridge);
+    const attachmentsInBlockSettings = blockSettings.attachments || [];
+
+    const onAddAttachments = (assets: Asset[]) => {
+        const newAssetIds = [];
+        for (const asset of assets) {
+            newAssetIds.push(asset.id);
+        }
+        setBlockSettings({
+            attachments: [...attachmentsInBlockSettings, ...newAssetIds.map((id) => ({ id }))],
+        });
+        addAssetIdsToKey(ATTACHMENTS_SETTING_ID, newAssetIds);
+    };
+    const onAttachmentDelete = (assetToDelete: Asset) => {
+        setBlockSettings({
+            attachments: attachmentsInBlockSettings.filter((asset) => asset.id !== assetToDelete.id),
+        });
+        deleteAssetIdsFromKey(ATTACHMENTS_SETTING_ID, [assetToDelete.id]);
+    };
+    const onAttachmentReplace = (attachmentToReplace: Asset, newAsset: Asset) => {
+        const indexOfReplacedAttachment = attachmentsInBlockSettings.findIndex(
+            (attachment) => attachment.id === attachmentToReplace.id
+        );
+        const attachmentsAfterReplace = [...attachmentsInBlockSettings];
+        attachmentsAfterReplace[indexOfReplacedAttachment] = { id: newAsset.id };
+        setBlockSettings({
+            attachments: attachmentsAfterReplace,
+        });
+        const newAssetIds = [];
+        for (const attachment of attachments) {
+            if (attachment.id !== attachmentToReplace.id) {
+                newAssetIds.push(attachment.id);
+            }
+        }
+        newAssetIds.push(newAsset.id);
+        updateAssetIdsFromKey(ATTACHMENTS_SETTING_ID, newAssetIds);
+    };
+
+    const onAttachmentsSorted = (assets: Asset[]) => {
+        const newAssetIds = [];
+        for (const asset of assets) {
+            newAssetIds.push(asset.id);
+        }
```

Comment:
same as above: could use a mapper. Or you could already pass the ids to the function `onAttachmentsSorted = (ids: string[])`

Reply:
The Asset[] is a special array with a custom iterator so mapping is not available unfortunately